### ISSUE TITLE
[BUG] Filtering a `ChoiceMap` with a `Selection` still allows the value to be accessed

### ISSUE THREAD

**Issue by ahiser1117** [genjax-community/genjax#1119]
**Describe the bug**
Using a very basic `ChoiceMap` I am able to access the value of an address that has been filtered out of the `ChoiceMap`

**To Reproduce**

Make a basic `GenerativeFunction`.
```python
from genjax import gen
from genjax import normal
from genjax import SelectionBuilder as S

@gen
def model():
    _ = normal(0.0, 1.0) @ "x"
    _ = normal(0.0, 1.0) @ "y"

key, subkey = jrand.split(key)
tr = model.simulate(subkey, ())
```

Get it's `ChoiceMap` and we can access info from it.
```python
normal_cmap: ChoiceMap = tr.get_sample()
normal_cmap['x']
```
```
Array(0.4807153, dtype=float32)
```
```python
'x' in normal_cmap, 'y' in normal_cmap
```
```
(True, True)
```

Filter the `ChoiceMap` with a complementary `Selection` and the value is still accessible.
```python
filtered_cmap: ChoiceMap = normal_cmap.filter(~S['x'])
filtered_cmap['x']
```
```
Array(0.4807153, dtype=float32)
```
Checking the existence says otherwise.
```python
'x' in filtered_cmap, 'y' in filtered_cmap
```
```
(False, True)
```

Filter using a normal `Selection` and try reading the value of the other address. Value is still accessed.
```python
other_filtered_cmap: ChoiceMap = normal_cmap.filter(S['x'])
other_filtered_cmap['y']
```
```
Array(-0.48219416, dtype=float32)
```

Again existence disagrees.
```python
'x' in other_filtered_cmap, 'y' in other_filtered_cmap
```
```
(True, False)
```

**Expected behavior**
I expect to get a `None` value or an error for accessing a missing address. When running,
```python
filtered_cmap['x']
```
```
None
```
or
```
Error
```

**Execution environment (please complete the following information):**
 - OS: Ubuntu 22.04
 - Python version: 3.12.3
 - GenJAX version: 0.4.0 Latest main commit as of 21:42 5/22 (994fb416a4d84ba22344648f076a7ab68b0767f3)

**Comment by ahiser1117** [comment]
In a similar vein. When I run,

```python
grad, nograd = Pytree.tree_grad_split(filtered_cmap)
```

`grad` still includes the filtered address. This becomes an issue for me when calling `NUTS` from `blackjax`.

I believe I can get the rest of `NUTS` working if this is remedied.

**Comment by ahiser1117** [comment]
I have just seen the current implementation of `__getitem__` for a `ChoiceMap` and the `TODO` that is written there.

**Comment by femtomc** [comment]
@ahiser1117 I made this behavior more clear in #1118 — instead of unwrapping masks, without the error (this could cause the issue you ran into)

**Comment by femtomc** [comment]
> grad still includes the filtered address. This becomes an issue for me when calling NUTS from blackjax.

This will still be true after #1118 -- but you won't be able to access the values as actual valid values (and therefore, in any inference or modeling code, it should be impossible _to use_ the value in a way which is non-differentiable).

I'm reasoning about this in a very loose way, I'm really interested to hear how it works in practice.

**Comment by ahiser1117** [comment]
@femtomc I made a dirty work around for it at the moment so I could move forward. Also interested to how this fairs in practice once this PR is merged.

**Comment by ahiser1117** [comment]
> This will still be true after #1118 -- but you won't be able to access the values as actual valid values (and therefore, in any inference or modeling code, it should be impossible _to use_ the value in a way which is non-differentiable).
> 
> I'm reasoning about this in a very loose way, I'm really interested to hear how it works in practice.

AFAIK this does not quite work the way that I need it to out of the box. My hope is to be able to do something along the lines of:
```python
filtered_chm = target.filter_to_unconstrained(tr.get_sample())
grad, nograd = Pytree.tree_grad_split((filtered_chm, tr.get_args()))

jtu.tree_flatten(grad)[0]
```
And get the following output (Expectation)
```python
[
<jax.Array(-0.30720285, dtype=float32)>,
<jax.Array(0.2336315, dtype=float32)>,
<jax.Array(0.17698473, dtype=float32)>,
<jax.Array(1.0572706, dtype=float32)>,
<jax.Array(0.9579706, dtype=float32)>,
<jax.Array(2.3612263, dtype=float32)>,
<jax.Array(-0.7195381, dtype=float32)>,
<jax.Array float32(38,) ≈0.12 ±1.0 [≥-2.1, ≤3.0] nonzero:38>,
<jax.Array float32(38,) ≈0.088 ±0.89 [≥-1.9, ≤1.8] nonzero:38>,
<jax.Array float32(38,) ≈-0.15 ±1.0 [≥-2.4, ≤2.4] nonzero:38>,
<jax.Array float32(38,) ≈-0.061 ±0.66 [≥-1.1, ≤1.5] nonzero:38>,
]
```
The output that I actually get is, (Reality)
```python
[
<jax.Array float32(38, 1600) ≈0.0024 ±1.0 [≥-2.6, ≤6.9] nonzero:60_800>,
<jax.Array(-0.30720285, dtype=float32)>,
<jax.Array(0.2336315, dtype=float32)>,
<jax.Array(0.17698473, dtype=float32)>,
<jax.Array(1.0572706, dtype=float32)>,
<jax.Array(0.9579706, dtype=float32)>,
<jax.Array(2.3612263, dtype=float32)>,
<jax.Array(-0.7195381, dtype=float32)>,
<jax.Array float32(38,) ≈0.12 ±1.0 [≥-2.1, ≤3.0] nonzero:38>,
<jax.Array float32(38,) ≈0.088 ±0.89 [≥-1.9, ≤1.8] nonzero:38>,
<jax.Array float32(38,) ≈-0.15 ±1.0 [≥-2.4, ≤2.4] nonzero:38>,
<jax.Array float32(38,) ≈-0.061 ±0.66 [≥-1.1, ≤1.5] nonzero:38>,
<jax.Array float32(38, 1600) ≈0.0024 ±1.0 [≥-2.6, ≤6.9] nonzero:60_800>,
]
```

The two 38x1600 arrays are expected to be removed by 
```python
filtered_chm = target.filter_to_unconstrained(tr.get_sample())
```
Since they're contraints in the target. 

Am I doing something wrong here / is there another way to do it? Currently I've defined my own `tree_grad_split` function that checks if the dimension of the array is 2 or not. Which is horrible and obviously not ideal. I'm cringing just describing it.

**Comment by femtomc** [comment]
The problem now with `Pytree.tree_grad_split` on `0.4.0` is that _inclusion_ of addresses (in filtered choice maps) is not directly reflected in the actual `Pytree` representation of the structure. 

Inclusion is computed correctly via the choice map interfaces, but before, the reason why `Pytree.tree_grad_split` worked is because the actual structure itself did not possess a pointer to the data for addresses which are not contained in the choice map.

Really, it wasn't a "proper" (meaning: via a well-defined Gen interface) way to do this before -- because `Pytree.tree_grad_split` is _beyond_ the interfaces -- it introspects on the representation itself, and sort of ignores the fact that the thing is an object, with a particular type, and particular interfaces.

I'm not exactly sure how to solve this _right now_ - it doesn't seem impossible to solve, but I'll need to think more about it.

**Comment by femtomc** [comment]
@ahiser1117 One thing which this is pointing out is that maybe we do want to have a Gen-focused HMC implementation.

We're embracing several ways of expressing uncertainty over structure which, as far as I can tell from this issue, our interaction with BlackJAX will trip over (most of these ways of expressing runtime uncertainty can't be reified into a concrete, known at compile time Pytree _with just the relevant pieces_ - some of these ways attach _other arrays_ to things, and we won't be able to collapse those down to a Pytree which we pass over to BlackJAX, and expect it to work correctly)

**Comment by ahiser1117** [comment]
Is it possible to tree_map along the `ChoiceMap` and `Selection` to include the constrained addresses in the `nograd` output? The benefit for my use case is that only happens once so it doesn't need to be super efficient. The current `Pytree.tree_grad_zip` implementation is used in the loop and works well.

**Comment by sritchie** [comment]
@ahiser1117 does your workaround still work? Also, ping to @femtomc on this question.

**Comment by ahiser1117** [comment]
@sritchie For my particular use case, it does, but it's literally just a stupid boolean check that you'd never want implemented into GenJAX. If a more elegant solution is created, it would be greatly appreciated.

I'll be at the events this week and maybe we can talk about it.

**Comment by sritchie** [comment]
The original bug is now solved on `main`:

```python
In [13]: from genjax import ChoiceMapBuilder as C
In [14]: import genjax

In [15]: chm = C["x"].set(10) ^ C["y"].set(12)
In [16]: filtered = chm.filter(genjax.SelectionBuilder["x"])

In [17]: "x" in filtered, "y" in filtered
Out[17]: (True, False)

In [18]: "x" in chm, "y" in chm
Out[18]: (True, True)

In [19]: filtered["y"]
---------------------------------------------------------------------------
ChoiceMapNoValueAtAddress                 Traceback (most recent call last)
Cell In[19], line 1
----> 1 filtered["y"]

File <@beartype(genjax._src.core.generative.choice_map.ChoiceMap.__getitem__) at 0x12e0f99e0>:56, in __getitem__(__beartype_object_5067543232, __beartype_getrandbits, __beartype_get_violation, __beartype_conf, __beartype_func, *args, **kwargs)

File ~/code/python/genjax/src/genjax/_src/core/generative/choice_map.py:590, in ChoiceMap.__getitem__(self, addr)
    588 v = submap.get_value()
    589 if v is None:
--> 590     raise ChoiceMapNoValueAtAddress(addr)
    591 else:
    592     return v

ChoiceMapNoValueAtAddress: ('y',)
```

**Comment by sritchie** [comment]
@ahiser1117 , @femtomc , feel free to open another issue if you all decide that that other problem, of values staying around with a mask in front of them vs actually disappearing, is solvable.

**Comment by sritchie** [comment]
@ahiser1117 , as of https://github.com/probcomp/genjax/pull/1359 we’ve added

a simplify() method to choicemaps
an eager=True option to filter on choicemaps

that will attempt to eagerly push the filters down to the leaves, cutting out values that don’t pass the filters, just as you had requested in https://github.com/probcomp/genjax/issues/1119

Hopefully this is a better solution than what we had before!

**Comment by ahiser1117** [comment]
@sritchie Oh fantastic. I'll go test this.

**Comment by ahiser1117** [comment]
@sritchie Okay this an awesome feature but I think there might be an issue. This might be due to how I've setup my generative functions and that's the real issue.

A random sample from my model looks like this:

```python
chm
```
![image](https://github.com/user-attachments/assets/31a8e4a1-64d0-42ab-900d-d91342da82ad)

So there's some latents with ndim=0. Within 'local' the latents have ndim=1 and length 94. But the value I want to filter is 'ys' which has shape (94, 1600). This is what happens when I filter with respect to that.

```python
filtered = chm.filter(cmap.get_selection(), eager=True)
filtered
```
![image](https://github.com/user-attachments/assets/40b6a215-34b3-4048-a3dc-69a66ebdc324)

At first this looked fine but I realized the `Mask` is weird:
![image](https://github.com/user-attachments/assets/a8aaf33e-8dd1-40d8-bd50-6d5f28d9b91a)

The problem really presents itself when I try to filter with the complement which is what I really want to do:
```python
filtered = chm.filter(~cmap.get_selection(), eager=True)
filtered
```
![image](https://github.com/user-attachments/assets/fb663b75-9d1f-465a-8c0f-723a04a57044)

The `'local',...,'ys'` values are still there, and the `Mask` has the same but complemented issue.
![image](https://github.com/user-attachments/assets/4a9ed764-537e-4c65-ab5c-49f4b037bfd4)

So is this an issue with the filtering process or is there something I should do about the last dimension in `'local',...,'ys'` so that it's delt with?

**Comment by sritchie** [comment]
@ahiser1117 can you show us how you made `cmap`?

**Comment by ahiser1117** [comment]
@sritchie 

```python
cmap = jax.vmap(lambda idx, y: C["local", idx, "ys"].set(y))(jnp.arange(ys.shape[0]), ys)
```
Where `ys.shape = (94, 1600)`

This represents the observational data I'm fitting the model to.

**Comment by ahiser1117** [comment]
For more context, this is from this function that returns the `GenerativeFunction` with `std_beh` cached. I've been considering whether it would make more sense to pass `std_beh` as an argument to the generative function. `std_beh.shape = (94, 1600, 3)`

```python
def model_generator_120(std_beh):

    @gen
    def kernel(beh, c_vT, c_v, c_θh, c_P, s, α, σ_SE):
        ζ0 = normal(0.0, 1.0) @ "ζ0"
        b = normal(0.0, 1.0) @ "b"
        y0 = normal(0.0, 1.0) @ "y0"
        σ0_noise = normal(0.0, 1.0) @ "σ0_noise"
        σ_noise = σ_NOISE_MEAN * jnp.exp(σ0_noise * σ_NOISE_STD)

        ζ = compute_ζ(ζ0)

        z = noiseless_model(c_vT, c_v * ζ, c_θh * ζ, c_P * ζ, 0, s, b, beh)

        _ = lgssm(z, y0, α, σ_SE * ζ, σ_noise) @ "ys"

        return z

    @gen
    def model():
        c_vT = normal(0.0, 1.0) @ "c_vT"
        c_v = normal(0.0, 1.0) @ "c_v"
        c_θh = normal(0.0, 1.0) @ "c_θh"
        c_P = normal(0.0, 1.0) @ "c_P"

        s0 = normal(0.0, 1.0) @ "s0"

        α0 = normal(2.0, 1.0) @ "α0"
        σ0_SE = normal(0.0, 1.0) @ "σ0_SE"

        α = norm.cdf(α0)
        σ_SE = σ_SE_MEAN * jnp.exp(σ0_SE * σ_SE_STD)

        s = compute_s(s0)

        z = (
            kernel.vmap(in_axes=(0, None, None, None, None, None, None, None)
                        )(std_beh, c_vT, c_v, c_θh, c_P, s, α, σ_SE) @ "local"
        )

        return z

    return model
```

**Comment by sritchie** [comment]
you know, I have a strong suspicion that this is the same bug as https://github.com/probcomp/genjax/pull/1341.

Can you try either branch `sritchie/yoni_fix`, or version "0.6.1.post26.dev0+e70a9764" (currently publishing, might be a couple minutes)? I have a feeling that that will sort you out.

You don't need vmap, btw:

```python
cmap = C["local", jnp.arange(ys.shape[0]), "ys"].set(ys)
```

and when I'm done with #1355 you'll be able to write

```python
cmap = C["local",:,"ys"].set(ys)
```

**Comment by ahiser1117** [comment]
@sritchie It appears this does not work. Although perusing through #1341 that looks very similar to this issue, so idk.

To confirm:
```python
print(genjax.__version__)
# 0.6.1.post26.dev0+e70a9764
```

The outputs from all of the above were the same.

**Comment by sritchie** [comment]
Sounds like we have some work to do. Thanks for this repro, Alex!

**Comment by ahiser1117** [comment]
@sritchie It seems I missed something! If I define the filtering `Selection` separately this ends up working.

```python
from genjax import SelectionBuilder as S

filter_sel = S["local",...,"ys"]
filter_sel
```
![image](https://github.com/user-attachments/assets/d4f97f09-92ea-4401-a872-4b022619105a)

```python
filtered = chm.filter(~filter_sel, eager=True)
filtered
```
![image](https://github.com/user-attachments/assets/f737c0a9-6c8a-43a0-ad7b-96197fa955d8)

What I was trying to do previously that wasn't working is taking the `Selection` from `cmap`.

```python
cmap.get_selection()
```
![image](https://github.com/user-attachments/assets/f578b5a3-68c2-4b9b-b95c-7ca7ecceed7d)

```python
filtered = chm.filter(~cmap.get_selection(), eager=True)
filtered
```

![image](https://github.com/user-attachments/assets/a597dea9-1b3a-49bd-b753-51d1ae432db9)

I guess it would be nice if I could do the latter, but I can make this work as is.

**Comment by sritchie** [comment]
I think this is definitely related to that other bug, and also to wonkiness with specifying indices… thanks again for the good report, I'll tackle this as part of our "make choicemaps nice" initiative.

**Comment by sritchie** [comment]
hey @ahiser1117 , if you can get me a smaller repro I think I have a fix for this at #1422 . Or if not, I understand why those masks were showing up well enough to fix this.

**Comment by sritchie** [comment]
@ahiser1117 I'm going to close this for now, BUT I would love a repro of that second issue you were hitting. it looks like there is some interaction with masks and vmap that I can now detect, but I would love your test case.

**Comment by ahiser1117** [comment]
@sritchie Just got out a meeting and saw this. I'll see if I can whip something up. I believe the issue just had to do with the `Indexed` dimension within `'local'` was shape (94,) but the values at `'local',...,'ys'` were shape (94,1600) where the second dimension was not `Indexed`. When attempting the filtering it was just trying to filter by the first dimension hence why the `Mask` had shape (94, 94).

**Comment by sritchie** [comment]
yup, excellent, that makes sense that the 94 dimension was created by an external vmap. I think we can auto-detect this and make the trace do the right thing and apply its own matching internal `jax.vmap`.

you could also handle this by vmapping what you were trying to do. This will drop down a level to where you have `()`and `(1600,)` shapes, which are compatible, and then `jax` will  sew the result back up correctly.

**Comment by ahiser1117** [comment]
Yeah that's a possibility.

Also, I just went through the same steps I had above and it works. Now I'll make a simpler repro so we can close this out officially!

**Comment by sritchie** [comment]
Thank you, you're the man

**Comment by ahiser1117** [comment]
Okay so this isn't super simple but I think it's about as simple as I can make it to fully encapsulate what I'm doing.

### Define simplified LGSSM definition (required to get that problematic 2nd dimension) and `GenerativeFunction` `model()`

```python
from genjax import gen, pretty, exact_density, normal, ChoiceMapBuilder as C
pretty()

def sample_step(carry, zt):
    key, a = carry
    key, subkey = jax.random.split(key)

    # Measurement update
    y_t = zt + a * jax.random.normal(subkey)

    return (key, a), y_t

def sample_lgssm(key, z, a) -> jnp.ndarray:
    carry = (key, a)

    _, y_sequence = jax.lax.scan(sample_step, carry, z)

    return y_sequence

def kalman_step(a, yt_zt):
    y_t, z_t = yt_zt

    innovation = y_t - z_t

    # Update log likelihood
    log_likelihood_t = -0.5 * (
        innovation**2 / a + jnp.log(2 * jnp.pi * a)
    )

    return a, log_likelihood_t


def logpdf_lgssm(y, z, a):
    _, log_likelihoods = jax.lax.scan(kalman_step, a, (y, z))

    # Sum log likelihoods to get total log likelihood
    total_log_likelihood = jnp.sum(log_likelihoods)

    return total_log_likelihood

def lgssm_dist():
    def sample(key, z, a):
        return sample_lgssm(key, z, a)

    def logpdf(y, z, a):
        return logpdf_lgssm(y, z, a)

    return exact_density(sample, logpdf)

lgssm = lgssm_dist()

def hier_model(vals_2d):

    @gen
    def kernel(vals_1d, a):
        ys = lgssm(vals_1d, a) @ "ys"
        
        return ys

    @gen
    def model():
        a = normal(0.0, 1.0) @ "a"

        z = (
            kernel.vmap(in_axes=(0, None))(vals_2d, a) @ "local"
        )

        return z

    return model
```

### Simulate model:

```python
key, subkey = jrand.split(key)
vec_2d = jrand.normal(subkey, (10, 100))

hm = hier_model(vec_2d)

key, subkey = jrand.split(key)
tr = hm.simulate(subkey, ())
chm = tr.get_sample()
chm
```
![image](https://github.com/user-attachments/assets/b3126fbc-5dab-4162-b17c-64b4619722f8)

### Generate some `ChoiceMap` as constraint that is then used to be filtered by:

```python
key, subkey = jrand.split(key)
cmap = C['local',:,'ys'].set(jrand.normal(subkey, (10, 100)))

sel = cmap.get_selection()
```

### Filter by `sel` and `~sel`

```python
chm.filter(sel, eager=True)
```
![image](https://github.com/user-attachments/assets/b7617ccc-ae9b-42d3-b24d-b62a0918e04f)

```python
chm.filter(~sel, eager=True)
```
![image](https://github.com/user-attachments/assets/01c0ee13-a855-46bd-9725-a31b3b11074d)
